src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -149,15 +149,31 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     // but ultimately represent the same underlying assembly. We compare the module version ids to determine if the references are the same
                     if (!compilationA.References.SequenceEqual(compilationB.References, new LambdaComparer<MetadataReference>((old, @new) => 
                     {
+                        if (ReferenceEquals(old, @new))
+                        {
+                            return true;
+                        }
+
                         if (old is null || @new is null)
                         {
                             return false;
                         }
 
-                        return compilationA.GetAssemblyOrModuleSymbol(old) is IAssemblySymbol oldAssembly
-                            && compilationB.GetAssemblyOrModuleSymbol(@new) is IAssemblySymbol newAssembly
-                            && oldAssembly.Modules.Select(m => m.GetMetadata()?.GetModuleVersionId() ?? Guid.Empty)
+                        var oldSymbol = compilationA.GetAssemblyOrModuleSymbol(old);
+                        var newSymbol = compilationB.GetAssemblyOrModuleSymbol(@new);
+
+                        if (oldSymbol?.Equals(newSymbol, SymbolEqualityComparer.Default) == true)
```

Comment:
I think you avoid the awkward nullable business like so.

```suggestion
                        if (SymbolEqualityComparer.Default.Equals(oldSymbol, newSymbol))
```

Reply:
I'm surprised this wasn't hit by an analyzer. Do we not have those enabled in our repo?